ocaml/xapi/xapi_pvs_cache_vdi.ml

```diff
@@ -0,0 +1,56 @@
+(*
+ * Copyright (C) 2016 Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open Client
+
+let size = Int64.of_int (1024 * 1024 * 1024) (* TODO: how big? *)
+
+let find_all ~__context ~sr =
+	let open Db_filter_types in
+	Db.VDI.get_refs_where ~__context ~expr:(And
+		(Eq (Field "SR", Literal (Ref.string_of sr)),
+		Eq (Field "type", Literal "pvs_cache")))
+
+let find_one ~__context ~sr =
+	match find_all ~__context ~sr with
+	| [] -> None
+	| vdi :: _ -> Some vdi
+
+let find_or_create ~__context ~sr =
+	match find_one ~__context ~sr with
+	| Some vdi -> vdi
+	| None ->
+		Helpers.call_api_functions ~__context (fun rpc session_id ->
+			Client.VDI.create ~rpc ~session_id
+				~name_label:"PVS cache VDI"
+				~name_description:"PVS cache VDI"
+				~sR:sr
+				~virtual_size:size
+				~_type:`pvs_cache
+				~sharable:false
+				~read_only:false
+				~other_config:[]
+				~xenstore_data:[]
+				~sm_config:[]
+				~tags:[])
+
+let destroy ~__context ~sr =
+	match find_all ~__context ~sr with
```

Comment:
I guess you pattern match the empty list to save a `Helpers.call_api_functions` call in that case?